### ISSUE TITLE
How to use Kinto for standard user (non-admin) on Linux?

### ISSUE THREAD

**Issue by rvbgnu** [rbreaves/kinto#629]
**First: Thank you very much for providing such a great and easy to install solution!** 

I am using Ubuntu 20.04 on a MacBook Pro, using both the internal keyboard, and an extended external one (UK version, with f13 to f19). I only installed the minimum Ubuntu, and I move step by step to get my workflow with Open Source / Free Software similar to what I had on Mac OS.

I successfully installed kinto.sh on my admin user.
I've searched your README file, and the open issues. I had to change the xkeysnail service settings with `sudo nano /lib/systemd/system/xkeysnail.service` with the following line:
```
#Restart=on-failure
Restart=always
```
so when I reboot or logout from the admin user, Kinto is running correctly.

But for my standard user, it is not available. I appreciate that the install on kinto is on the admin user side, so it is not a surprise. With my little system admin skills, I cannot find a way to make it work for all uses of the Linux install.

Note: I kept the habitude, from using Mac OS, and when I have to work on Windows, to have a standard user for every day work, for each person, and to only use the admin/sudo user for maintenance, installs and updates. I find it safer, isolated, and necessary for privacy concerns.

Please could you help me to make Kinto work for a standard user? I will be glad to test and help writing the necessary explanation for other user, like me, who wants to achieve that.

**Comment by Atemu** [comment]
Does xkeysnail need privileges? Otherwise, a systemd user service could be used instead.

**Comment by rvbgnu** [comment]
Thank you for your reply.
I am not familiar with systemd user service and underlying system tools, I'm sorry. I switched from Mac desktop.
After reading more about xkeysnail, it seems that  

> - Cons
>    Runs in root-mode (requires sudo)

https://github.com/mooz/xkeysnail#xkeysnail

**Comment by Atemu** [comment]
Well, there you have it: No xkeysnail without superuser privileges.

**Comment by rbreaves** [comment]
That’s not exactly true - there is a method for assigning uinput to a non-sudo level privilege. How to do this has been discussed btwn me, redbearAK & Lenbok - it just has not been implemented in Kinto just yet.

It is something I’m interested in doing - BUT I need or want to reconfigure several distros for testing before making a change like that. (Moving away from virtualbox)

The conversation already exists - just too lazy to link it right now. I’m closing this as a duplicate though.

**Comment by rvbgnu** [comment]
All right, thank you very much for your reply. I am impressed by your commitment and swift response.

I would prefer not to work everyday as a super user, I am not experienced enough with Ubuntu. And my priority is to switch and move data and most of my apps or alternative from Mac, and I already made some mistakes, so it is safer for me at the moment.

However, I wonder if launching kinto.sh with privileges would be enough. Something like when the Snap/flatpack app request an administrative password. I wouldn't mind doing that each time I log my user session.

I will look around in the GitHub issue about the conversation you mentioned before.

**Comment by rvbgnu** [comment]
Maybe here
https://github.com/rbreaves/kinto/issues/609
or the mention inside
https://github.com/mooz/xkeysnail/issues/64#issuecomment-600380800

I will be more than happy to test on "vanilla user" that I can create on my current Ubuntu install 20.04 LTS. I also use long aluminium external Apple keyboard, with arrow, numeric pad and F13 to F19 functions keys. So please fill free to ping me.

So in the meantime, I have to read back my previous searches before finding your beautiful kinto.sh 😓
at least to get most Mac shortcuts Cmd+C V or X shortcut, for my muscle memory! I am not an avid Terminal user, so I can compromise 😜
Please would you point me to (less good, I know) alternatives?

**Comment by rbreaves** [comment]
Hmmm point you to what other type of alternatives? macOS like remappers that don't require sudo access? I don't have a list at the read but there is this made for autokey https://github.com/petrstepanov/gnome-macos-remap.

Also my earlier version of Kinto used a custom C program to detect what app was in focus (x11) and would use xkb to remap. It was far too difficult to update, maintain and expect others to contribute much too and had its issues. Although if you have to use something like xRDP for linux then it may be one of the only options. Why I switched to xkeysnail, & better alt-tab type remapping & many improvements really.
https://github.com/rbreaves/kinto/releases/tag/1.0.7-3

**Comment by rbreaves** [comment]
Also you can likely set the uinput to a non-sudo user manually and then just edit the xkeysnail service to use your user profile I assume and remove the reference to sudo in the xkeysnail.service file (relocated to your user profile). All things I will have to automate and test though before merging into master.

**Comment by rvbgnu** [comment]
All right, it sounds like a good work around. I will have to look a bit more into the succession of commands. I will do my best. Would it be the right place to report back here, in the comments, or where else? Then it could be reproduced by anyone who wants to install and test.

I understand that you want to automate and test before merging. A partly manual setup will be good enough for me, and maybe other users!!
It is like the switch user issue that I read here, it is very rare, and not that difficult to work around: log out, and log in again (well, I wish I had the "saved state and re-open at login options like on Mac, but we can't have it all !!)

**Comment by rbreaves** [comment]
Absolutely, if you have any questions or issues trying what I suggest just let me know here. Normally I try to provide better guidance, but since I haven't gone through the entire setup myself it is hard to say for sure what configuration will work for a system level systemd service.

Although having said that the original Kinto app based on xkb and my custom C was indeed a user level systemd service. So I have done it before, just not with xkeysnail/uinput. I assume it will work fairly easily under most distros, particularly Ubuntu, but Fedora is a question mark for me as even getting the sudo level service working on that one was more difficult than most due to selinux.

**Comment by rvbgnu** [comment]
Thank you Ben, appreciate it. So I give it a try, I mean several tries (and I also love rugby :rugby_football: so I won't stop after one try ;-)
I am a bit out of my depth with Linux system admin, but I search for information as much as possible.

> 1. set the uinput to a non-sudo user manually

The little I found about this seems to be very un-secure, and just changing the permissions did not help.

> 2. remove the reference to sudo in the xkeysnail.service file (relocated to your user profile)

2.1. Disable the systemd service on the root user (my initial installation, that work well with a root user)
2.2. Copy and run the systemd service at standard user level (in ~/.config/systemd/user/xkeysnail.service).  
I've checked that the service is running, and rebooted as well, all good: service only the standard user.

> 3. edit the xkeysnail service to use your user profile

3.1. Copy the kinto script with gui folder on the standard user
```~/.config$ sudo cp -r kinto/ /home/[standard-user]/.config/.```
3.2 Editions of the script:
```
# @test for user non-root: principal, instead of roothb
#ExecStart=/usr/bin/sudo /bin/bash -c '/usr/bin/xhost +SI:localuser:root && /home/[root-user]/.config/kinto/killdups.sh && /usr/local/bin/xkeysnail --quiet --watch /home/[root-user]/.config/kinto/kinto.py'
ExecStart=/bin/bash -c '/usr/bin/xhost +SI:localuser:[standard-user] && /home/[standard-user]/.config/kinto/killdups.sh && /usr/local/bin/xkeysnail --quiet --watch /home/[standard-user]/.config/kinto/kinto.py'
#ExecStop=/usr/bin/sudo /bin/bash -c '/usr/bin/sudo pkill -f bin/xkeysnail && exit 0'
ExecStop=/bin/bash -c 'pkill -f bin/xkeysnail && exit 0'

```

For now it seems that there are pieces missing on the standard user side. I can run kinto.sh
```~/.config/kinto/gui/kinto-gui.py```
 but I have errors in the logs:
```
xkeysnail:     $ sudo xkeysnail config.py
Make sure that you have executed xkeysnail with root privilege such as  
Failed to open `uinput` in write mode. 
                   v0.3.0

```

So I will try again, I probably have to make edits in the `~/.config/kinto/gui/kinto-gui.py` file as well.

As I cannot install from the `set-up.py` file, abviously for the [standard-user], what is missing here, that I could copy from the root user?

**Comment by rvbgnu** [comment]
So I am trying to get a ExecStart command that works, but to no avail yet - [standard-user] stand for my user name here:
```
$ /bin/bash -c '/usr/bin/xhost +SI:localuser:[standard-user] && /home/[standard-user]/.config/kinto/killdups.sh && /usr/bin/sudo /usr/local/bin/xkeysnail --quiet --watch /home/[standard-user]/.config/kinto/kinto.py'

localuser:[standard-user] being added to access control list
[sudo] password for [standard-user]:

```
attemps failed because [standard-user] is not a sudoer.

**Comment by rbreaves** [comment]
> > 1. set the uinput to a non-sudo user manually
> 
> The little I found about this seems to be very un-secure, and just changing the permissions did not help.

I do agree with this assessment, if I am to be honest, but I do not know of any other way as xkeysnail depends on uinput and uinput either needs sudo access or to be added to another user/group so that it can run.. but I do agree that making that sort of change could have unintended consequences.. if someone knew your uinput was accessible as a non-sudo user I am not entirely sure of the implications. At least with sudo you have to elevate with a password explicitly or add the command to sudoers.

Perhaps the better option is to keep the sudoer approach, but also find a way to run the xkeysnail process itself as a local user using local systemd services while having it only escalating the calls to uinput as a sudo type user, however it or its libraries do that. Not sure what that would like it but just a thought.

**Comment by rvbgnu** [comment]
good, thank you. I think it is the right direction. I will go it more tries when I have some ideas or inspiration;-)

For now, I used your alternative from an earlier version
> Also my earlier version of Kinto used a custom C program to detect what app was in focus (x11) and would use xkb to remap. [...] Although if you have to use something like xRDP for linux then it may be one of the only options.
https://github.com/rbreaves/kinto/releases/tag/1.0.7-3

I don't need RDP or VNC really, so both ways is fine for me. 
This alternative works well, it has stopped twice for no reason or error in the logs, but that's ok. Logout and login and it is back. I executed the setup.py script on my standard user, and I ran the sudo commands in another Terminal window in parallel, with the root user. I had to run setup a few times, to cover my mistakes or misunderstandings.

A few missing shortcuts, like switching tabs in windows, for Firefox/Thunderbird, it is `Cmd + PgUp / PgDown` instead of `Ctrl + Tab`, and Ubuntu doesn't provide remapping of that in the `Settings > Keyboard Shortcuts`. But at least `Ctrl + Tab` is doing nothing else, so it doesn't disturb me that much. But I could remap in `Settings` the `Cmd + Tab` to switch Applications, so it is good enough for my workflow now.

Thank you again for providing guidance and help to get a solution!
Once I have another computer, custom build for Linux (and ethical, eco-friendly) the keyboard will not be a Mac, so Super key should be close to Space bar, and less adjustments for my muscle memory. As I expect to work on different computers and setup, it should streamline my keyboard shortcuts across the different systems. This is a huge time and energy saver!
